generate interface files

OceanOak opened the pull request:
Generating interface files (.resi ) for all implementation files (.res)

pbiggar commented on src/tea_html.resi:
```diff
@@ -0,0 +1,203 @@
+module Cmds = Tea_html_cmds
+let map: ('a => 'b, Vdom.t<'a>) => Vdom.t<'b>
+let noNode: Vdom.t<'a>
+let text: string => Vdom.t<'a>
```
Can you add some spacing to split them into rough sections? Use your judgment for what should be grouped together in the sections.

pbiggar replied:
And the same for the other files.

pbiggar reviewed (changes requested):
Great, I think we can merge this with some changes.

You should be able to see the decoder types here that we want to change to rescript-json-combinator types for #3.

pbiggar commented on src/tea_html.res:
```diff
@@ -1,6 +1,6 @@
 open Vdom
 
-module Cmds = Tea_html_cmds
+//module Cmds = Tea_html_cmds
```
If these are being removed, can you make sure to remove them, not just comment them out

OceanOak replied:
Yes, I just wanted you to see it before completely removing them .

OceanOak commented:
**Changes made** 

**tea_json.resi**

**1**

`    type rec t<'a> = Map.Make(String).t<'a>
`
Map.Make didn't work inside .resi file used Belt.Map.String instead in both .resi and res file to match.

**2**

Used Js.Json.parseExn(string) instead of Web.Json.parseExn(string) because of The value parseExn can't be found in Web.Json error

**3**

removed some values from ObjectDict because of this error Expected declaration In module Decoder.ObjectDict: The value `empty' is required but not provided. 

example on what  was removed
```
    let empty: t<'a>
    let is_empty: t<'a> => bool
    let mem: (key, t<'a>) => bool
```

**tea_svg.resi**

`let switch: (~key: string=?, ~unique: string=?, Vdom.properties<'a>, list<Vdom.t<'a>>) => Vdom.t<'a>
`

switch is a reserved keyword error. Changed it to:

`let \"switch": (~key: string=?, ~unique: string=?, Vdom.properties<'a>, list<Vdom.t<'a>>) => Vdom.t<'a>
`

**tea.res**
Didn't generate .resi file for tea.res, it only containes modules.


All the rest of the files were generated without any errors.

pbiggar commented:
> **Changes made**
> 
> **tea_json.resi**
> 
> **1**
> 
> ` type rec t<'a> = Map.Make(String).t<'a>` Map.Make didn't work inside .resi file used Belt.Map.String instead in both .resi and res file to match.

This is a change we should make anyway, so this is fine.

> **2**
> 
> Used Js.Json.parseExn(string) instead of Web.Json.parseExn(string) because of The value parseExn can't be found in Web.Json error

The issue here is that the new file Web_json.resi only has definitions for things defined in Web_json.res. However, Web_json.res also `include`s Js.Json. So we accidentally hide the Js.Json stuff by adding the .resi. I think the best approach here would be to remove Web_json.resi for now, as it will break things for other users.

This change is fine though.


> **3**
> 
> removed some values from ObjectDict because of this error Expected declaration In module Decoder.ObjectDict: The value `empty' is required but not provided.
> 
> example on what was removed
> 
> ```
>     let empty: t<'a>
>     let is_empty: t<'a> => bool
>     let mem: (key, t<'a>) => bool
> ```
> 
> **tea_svg.resi**
> 
> `let switch: (~key: string=?, ~unique: string=?, Vdom.properties<'a>, list<Vdom.t<'a>>) => Vdom.t<'a> `
> 
> switch is a reserved keyword error. Changed it to:
> 
> `let \"switch": (~key: string=?, ~unique: string=?, Vdom.properties<'a>, list<Vdom.t<'a>>) => Vdom.t<'a> `

Can you rename this function `switch'` instead?

pbiggar commented:
Great, thanks!
